### ISSUE TITLE
Combination of stream decoders doesn't emit all expected items

### ISSUE THREAD

**Issue by mpod** [scodec/scodec-stream#169]
Output of the following program is `abcd`, while according to my understanding it should be `abcdababc`. Am I missing something?

```scala
import cats.effect.{ExitCode, IO, IOApp}
import fs2.Stream
import scodec.Err
import scodec.bits.{BitVector, HexStringSyntax}
import scodec.codecs.bits
import scodec.stream.StreamDecoder

object Test extends IOApp {
  val input = hex"1a bc d7 ab 7a bc"

  val decoder = StreamDecoder.many(bits(4)).flatMap { a =>
    //println(s"a: ${a.toHex}")
    StreamDecoder.tryMany(
      bits(4).flatMap { b =>
        //println(s"b: ${b.toHex}")
        if (b.toHex == "7") scodec.codecs.fail[BitVector](Err("AAA"))
        else scodec.codecs.provide(b)
      }
    )
  }

  override def run(args: List[String]): IO[ExitCode] =
    Stream
      .emits(input.toArray.map(BitVector.apply(_)))
      .covary[IO]
      .through(decoder.toPipe)
      .evalTap(a => IO(println(s"Decoded: ${a.toHex}")))
      .compile
      .drain >> IO(ExitCode.Success)
}
```

**Comment by mpilquist** [comment]
What version of scodec-stream are you using? Using 2.x or 3.x gives the expected behavior:

```scala
scala> decoder.decode[Fallible](Stream(hex"1a bc d7 ab 7a bc".bits)).compile.fold(BitVector.empty)(_ ++ _)                                                                                   
val res7: Either[Throwable, scodec.bits.BitVector] = Right(BitVector(36 bits, 0xabcdababc))
```

**Comment by mpod** [comment]
I don't see how your example is equivalent to mine.

Here is my `build.sbt`:

```scala
scalaVersion := "2.13.6"

libraryDependencies ++= Seq(
  "org.scodec"                 %% "scodec-stream"   % "3.0.1",
)
```

`$ sbt run` returns:
```
... 
Decoded: a
Decoded: b
Decoded: c
Decoded: d
```

**Comment by mpilquist** [comment]
Okay yeah, I can reproduce your results if I chunk up the input in to an individual `BitVector` for each byte of input:

```scala
scala> decoder.decode[Fallible](Stream.emits(hex"1a bc d7 ab 7a bc".toArray.map(BitVector.apply(_)))).compile.fold(BitVector.empty)(_ ++ _)                                                                                          
val res9: Either[Throwable, scodec.bits.BitVector] = Right(BitVector(16 bits, 0xabcd))

**Comment by mpilquist** [comment]
Fixed in 3.0.2 and 2.0.3, which are releasing now and should be available for download within the hour.
